[#356] Crash verify backup while save state

HonzaR opened the pull request:
This PR fixes crashing the app while in backup flow. Tested with different ways of configuration changes (e.g. kill the app in background, change theme/orientation, etc.). But there is one thing, which needs to be discussed first - do we want to introduce Parcelable to our code? As I know its not directly supported by Kotlin Multiplatform. But it can be done like described for example here in this [post](https://www.droidcon.com/2021/11/29/android-parcelable-in-kotlin-multiplatform/).

Closes https://github.com/zcash/secant-android-wallet/issues/356, https://github.com/zcash/secant-android-wallet/issues/246.

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author
<!-- NOTE: Do not modify these when initially opening the pull request.  This is a checklist template that you tick off AFTER the pull request is created. -->
- [x] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [x] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Manual tests: Did you update the [manual tests](../blob/main/docs/testing/manual_testing) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed._ Code Coverage can be run with: `./gradlew connectedCheck -PIS_COVERAGE_ENABLED=true`
- [ ] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes?
- [ ] Screenshots: Did you provide before and after UI screenshots in the description of this pull request?  _This is only applicable for changes that modify the UI._
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._


# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [x] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after changes and when possible, leave it in a better place._
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [x] Run the app: Did you run the app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

ccjernigan reviewed (dismissed):
I've been avoiding Parcelable and Parcelize so far, specifically because it makes a Multiplatform transition more difficult in the future.

In some cases, I've written a custom `Saver` implementation.  But I don't know if we need that in this case.

I think what would really help is to identify to where is Index being used as a directly parceled object or a field in a parceled object?  Once we know where the usage is, we can figure out a good solution.

ccjernigan reviewed (commented):
I've been avoiding Parcelable and Parcelize so far, specifically because it makes a Multiplatform transition more difficult in the future.

In some cases, I've written a custom Saver implementation. But I don't know if we need that in this case.

I think what would really help is to identify to where is Index being used as a directly parceled object or a field in a parceled object? Once we know where the usage is, we can figure out a good solution.

HonzaR commented:
> I've been avoiding Parcelable and Parcelize so far, specifically because it makes a Multiplatform transition more difficult in the future.
> 
> In some cases, I've written a custom Saver implementation. But I don't know if we need that in this case.
> 
> I think what would really help is to identify to where is Index being used as a directly parceled object or a field in a parceled object? Once we know where the usage is, we can figure out a good solution.

Yes, understand. `Index` objects are used in the `TestChoices` object. These `Index` objects are also a part of saved and restored state with _SavedStateHandle_ in the `BackupViewModel.kt`. There is `init{}` block, which saves map of `Index - String` pairs to `savedStateHandle` under the `KEY_TEST_CHOICES` key.

ccjernigan commented:
> > I've been avoiding Parcelable and Parcelize so far, specifically because it makes a Multiplatform transition more difficult in the future.
> > In some cases, I've written a custom Saver implementation. But I don't know if we need that in this case.
> > I think what would really help is to identify to where is Index being used as a directly parceled object or a field in a parceled object? Once we know where the usage is, we can figure out a good solution.
> 
> Yes, understand. `Index` objects are used in the `TestChoices` object. These `Index` objects are also a part of saved and restored state with _SavedStateHandle_ in the `BackupViewModel.kt`. There is `init{}` block, which saves map of `Index - String` pairs to `savedStateHandle` under the `KEY_TEST_CHOICES` key.

OK, cool. So it seems that maybe we have two options?
 - Tweak BackupViewModel to have a layer of indirection, where we we convert Index to Int, then back to Index for the purpose of shoving into the `savedStateHandle`
 - Similar to #383, move the state we want retain out of ViewModel and into a `rememberSaveable`.  At the same time, implement a custom `Saver`

Thoughts on those two?

HonzaR commented:
> > > I've been avoiding Parcelable and Parcelize so far, specifically because it makes a Multiplatform transition more difficult in the future.
> > > In some cases, I've written a custom Saver implementation. But I don't know if we need that in this case.
> > > I think what would really help is to identify to where is Index being used as a directly parceled object or a field in a parceled object? Once we know where the usage is, we can figure out a good solution.
> > 
> > 
> > Yes, understand. `Index` objects are used in the `TestChoices` object. These `Index` objects are also a part of saved and restored state with _SavedStateHandle_ in the `BackupViewModel.kt`. There is `init{}` block, which saves map of `Index - String` pairs to `savedStateHandle` under the `KEY_TEST_CHOICES` key.
> 
> OK, cool. So it seems that maybe we have two options?
> 
> * Tweak BackupViewModel to have a layer of indirection, where we we convert Index to Int, then back to Index for the purpose of shoving into the `savedStateHandle`
> * Similar to [Onboarding - refactoring of UI state retention into rememberSaveable fields #383](https://github.com/zcash/secant-android-wallet/issues/383), move the state we want retain out of ViewModel and into a `rememberSaveable`.  At the same time, implement a custom `Saver`
> 
> Thoughts on those two?

Super. Although the first solution will probably be easier to implement, I do like the second option more, because it will be more consistent with the rest of the code and as you mentioned, there is another issue with similar requirements. Do you agree with the second one?

ccjernigan commented:
> 

Agree, I also like the 2nd option a lot.

HonzaR commented:
Changes:

- Removed `Parcelable` from `Index` class.
- Removed `testChoices` from `BackupViewModel`.
- Added custom Saver class for `TestChoices`.
- Added `TestChoicesTest` and `TestChoicesFixture` classes for testing the saver.
- Added `selectedTestChoices` as rememberSaveable and changed `BackupWallet` api to be able to test it.
- Added integration test classes `BackupIntegrationTest`, `TestBackupActivity`, `BackupTestSetup` for Backup screens.

Seems like there are a lot of changes. Most of them are related to integration tests, which could probably be moved out to the next issue. Apologies for that, it went too smoothly :) It follows the same patter as implemented in the Onboarding integration tests.

ccjernigan commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/AndroidBackup.kt:
```diff
@@ -0,0 +1,73 @@
+import co.electriccoin.zcash.ui.screen.backup.viewmodel.BackupViewModel
+
+@Composable
+internal fun MainActivity.WrapBackup(
```
No changes needed here.  Informational only.  

This extension function could potentially go away, it is really just syntactic sugar.  I added it for now, because I like to call WrapBackup from MainActivity without passing an explicit `this` reference.  Seeing that explicit `this` to an Activity catches my eye during code review as doing so can sometimes be a potential memory leak.

ccjernigan commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/AndroidBackup.kt:
```diff
@@ -0,0 +1,73 @@
+}
+
+// This layer of indirection allows for activity re-creation tests
+@Composable
```
No changes needed here.  Informational only.  

This method is basically only visible for integration testing purposes.

ccjernigan commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/AndroidBackup.kt:
```diff
@@ -0,0 +1,73 @@
+    persistableWallet: PersistableWallet,
+    onBackupComplete: () -> Unit
+) {
+    val backupViewModel by activity.viewModels<BackupViewModel>()
```
No change needed here.  I'm looking ahead at what I was thinking how this might evolve.

If we eliminate this ViewModel, then we can alter that pattern a bit by having two files:

AndroidBackup.kt - contains Android-specific injections (e.g. the copy to clipboard method).

StatefulBackup.kt - A new file.  Contains the state retention logic, then delegates to `BackupWallet`.  This ends up being a cross-platform file, since it no longer touches ViewModel.  `BackupWallet` remains "view only" which makes it easy to continue testing the basic view behavior.

HonzaR replied:
Nice look ahead, thank you for your clarification.

ccjernigan reviewed (approved):
This is a great improvement; thanks for the extra effort that went into updating the integration tests too.  I pushed two commits to fix minor bugs—please review and confirm those work as expected.

I pushed a third commit that contains a refactoring, which has a new pattern that I think we should consider for starting to move logic out of MainActivity.  I think doing so makes it easier to test, and reverses the trend of MainActivity continuing to grow.  We might tweak it a bit as we try it with other screens, but I think it is a positive improvement.  I'm going to add a few comments line to point out what I was thinking when I made this suggested change.

HonzaR commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/view/BackupView.kt:
```diff
@@ -57,13 +63,17 @@ fun ComposablePreview() {
     wallet: PersistableWallet,
     backupState: BackupState,
-    selectedTestChoices: TestChoices,
+    choices: TestChoices,
```
Yes, moving the rememberSaveable field up a level is better solution. Thank you.

HonzaR commented:
@ccjernigan thank you for the cooperation on setting up our integration tests. It will be very useful for other screens. We can merge these changes now.
